### ISSUE TITLE
Add a guide about gRPC server reflection protocol

### ISSUE THREAD

**Issue by remyleone** [grpc/grpc.io#103]

**Comment by muxi** [comment]
@srini100 @ejona86 - office hour ping

**Comment by ejona86** [comment]
@remyleone, what are you hoping for here? A guide that shows you how to enable reflection on the server and hit the server with grpcurl? Or more about the protocol itself and how it works, such that you could make your own client or some such?

**Comment by remyleone** [comment]
A little bit of both but at least a document that provides an overview of what it is and how to enable it. At the moment, the content that are present on subdomains of grpc.io are not easy to find from grpc.io: 

https://grpc.github.io/grpc/core/md_doc_server-reflection.html 
https://grpc.github.io/grpc/core/md_doc_server_reflection_tutorial.html
https://grpc.github.io/grpc/python/grpc_reflection.html
https://grpc.github.io/grpc/core/md_doc_command_line_tool.html

By the way, what is the relationship between grpc.io and grpc.github.io?

**Comment by chalin** [comment]
> ... what is the relationship between grpc.io and grpc.github.io?

grpc.github.io is the old site, which will be decommissioned ([#22294](https://github.com/grpc/grpc/issues/22294)) -- if all goes well, I'll likely have some time today to work on that.

**Comment by ejona86** [comment]
@chalin, those links to grpc.github.io are API reference docs.

I wasn't aware of those md_doc_ pages in the generated documentation. But they are just a rendered version of the content in https://github.com/grpc/grpc/tree/master/doc , which we do reference a reasonable amount. grpc.io has relatively little in terms of deep technical docs; it is mostly getting started and introductions. Deeper docs are split between that doc folder and https://github.com/grpc/proposal

**Comment by chalin** [comment]
> @chalin, those links to grpc.github.io are API reference docs.

Right, and all those pages will be left as is, though the old site doc pages will be removed or marked obsolete (https://github.com/grpc/grpc.github.io/pull/858).

> grpc.io has relatively little in terms of deep technical docs

Maybe it's time to get some more technical docs into grpc.io 😄
I believe this is what @remyleone is asking for, but I'll let him comment on that.

**Comment by remyleone** [comment]
I would like to have a single official website where I can find everything related to gRPC, so it would be: references, getting started documentation and technical documentation. In the case of reflection, that would be a getting started guide that will then point to the deeper technical documentation that explains how it is implemented.

**Comment by ejona86** [comment]
> Maybe it's time to get some more technical docs into grpc.io ￼

We purposefully don't, because grpc.io isn't the source-of-truth. So such things get out-of-date. Introduction guides and pointers to more information are all fair game though, since those age better. Reference documentation is much "closer" to developers, so tends to be more likely up-to-date. And everyone mostly agrees that reference documentation is important (although quality will still vary).

grpc.io could be successful as a curated/organized ground for pointing to other places. But it will take a concerted effort from all members of grpc and buy-in, as otherwise it will be consistently out-of-date. To give some idea, we have quite a lot of content at https://github.com/grpc/proposal . But it is time-ordered and some gRFCs replace/supplement earlier gRFCs and it is hard to know how they fit together.

It seems quite feasible to have a "reflection" page on grpc.io and it just links to the reference documentation in each language and the gRFC/doc (in this case [server-reflection.md](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md)) and .proto. That seems like something that would scale better and deals with the problem that some things are in grpc/doc and some are a gRFC(s).

(@chalin, I will note that there is some Google-internal documentation for this sort of thing that is mainly basic intro guides. I'd be really happy if it was exported to github, but it is also not quite trivial to do that because there's a lot of google-isms. It is mostly per-language guides. For example, Java's reflection guide (internal to Google at go/grpc/java/server_reflection) has a blurb about what reflection is, links to grpc cli, reflection.proto, and basically says "use io.grpc.protobuf.services.ProtoReflectionService".)

(This is all related to a dream of mine to have a feature vs language matrix which links to API/docs for each language along with links to design docs for that feature.)

**Comment by chalin** [comment]
@ejona86 - maybe we could meet with @thisisnotapril and discuss what it would take to migrate the internal docs to grpc.io?

**Comment by ejona86** [comment]
@chalin, that would be fabulous.

**Comment by chalin** [comment]
> I would like to have a single official website where I can find everything related to gRPC

@remyleone - we'll be making some headway into this wish: see #180, #182.
